[Logs] Adds Log Explorer docs (#13312)
cloudflare/cloudflare-docs: content/logs/log-explorer.md

{{<Aside type="note">}}

Log Explorer is currently in beta. To request access, complete the [sign-up form](https://cloudflare.com/lp/log-explorer/).

{{</Aside>}}

## Supported datasets

The following zone-level datasets are currently available with Log Explorer:

* [HTTP Requests](/logs/reference/log-fields/zone/http_requests/) (`FROM http_requests`)
* [Firewall Events](/logs/reference/log-fields/zone/firewall_events/) (`FROM firewall_events`)

## Authentication

In order to communicate with the API, you will need to configure the appropriate authentication headers.

* `X-Auth-Email` - the Cloudflare account email address associated with the domain
* `X-Auth-Key` - the Cloudflare API key

Alternatively, API tokens with Logs write permissions can also be used for authentication:

* `Authorization: Bearer <API_TOKEN>`

## Enabling Log Explorer

Use the Log Explorer API to enable Log Explorer for each dataset you wish to store. It may take up to 30 minutes after a logstream is enabled before logs are viewable.

The following curl command is an example for enabling `http_requests`, as well as the expected response when the command succeeds.

```bash

> curl --request POST \
>   --url https://api.cloudflare.com/client/v4/zones/<ZONE_ID>/logs/explorer/datasets \

Reviewer: ```suggestion
curl https://api.cloudflare.com/client/v4/zones/{zone_id}/logs/explorer/datasets \
```
POST is the default when there's `--data` in the request.

6. Select the **Time period** from which you want to query, for example, the previous 12 hours.

![Log explorer dashboard](/images/logs/log-explorer-dash.png)

7. Select **Add filter** to create your query. Select a **Field**, an **Operator**, and a **Value**.

![Log explorer filters](/images/logs/log-explorer-filters.png)

8. A query preview is displayed. Select **Use custom SQL**, if you would like to change it.
9. Select **Run query**, when you are done. The results are displayed below within the **Query results** section.

{{<Aside type="note">}}

You can also access the Log Explorer dashboard directly from the [Security Analytics dashboard](/waf/analytics/security-analytics/#logs). When doing so, the filters you applied in Security Analytics will automatically carry over to your query in Log Explorer.

{{</Aside>}}

## Use Log Explorer from the Query API

Log Explorer exposes an easy to use query endpoint that uses a familiar SQL syntax for querying your logs generated with Cloudflare's Network.

For example, to find a HTTP request with a specific RayID, you can perform the following SQL query.

```bash

> curl --request GET \
>   --url https://api.cloudflare.com/client/v4/zones/<ZONE_ID>/logs/explorer/query/sql \

Reviewer: ```suggestion
curl https://api.cloudflare.com/client/v4/zones/{zone_id}/logs/explorer/query/sql \
```
GET is the default when there's no data.

  --header "Authorization: Bearer <API_TOKEN>" \
  --url-query query="SELECT clientRequestScheme, clientRequestHost, clientRequestMethod, edgeResponseStatus, clientRequestUserAgent FROM http_requests WHERE RayID = '806c30a3cec56817' LIMIT 1"
```

Which returns the following HTTP Request details:

```bash
{
  "result": [
    {
      "clientrequestscheme": "https",
      "clientrequesthost": "example.com",
      "clientrequestmethod": "GET",
      "clientrequestuseragent": "curl/7.88.1",
      "edgeresponsestatus": 200
    }
  ],
  "success": true,
  "errors": [],
  "messages": []
}
```

## Output formats

> Log Explorer output can be presented in different formats, besides JSON: JSON Lines (also known as, NDJSON), CSV, and plain text. The plain text uses ASCII tables similar to psql's `aligned` output mode. Besides the convenience factor of not having to translate the format on the client side, JSON Lines, CSV and plain text formats have the advantage of being streamed from the API. So for large result sets, you will get a response earlier.

Reviewer: ```suggestion
Log Explorer output can be presented in different formats, besides JSON: JSON Lines (also known as NDJSON), CSV, and plain text. The plain text uses ASCII tables similar to psql's `aligned` output mode. Besides the convenience factor of not having to translate the format on the client side, JSON Lines, CSV, and plain text formats have the advantage of being streamed from the API. So for large result sets, you will get a response earlier.
```

You can choose the output format with an HTTP `Accept` header, as shown in the table below:

> | Output format | Content type | Streaming? |
> |---------------|------------------------|------------|
> | JSON | `application/json` | no |
> | JSON Lines | `application/x-ndjson` | yes |
> | CSV | `text/csv` | yes |
> | plain text | `text/plain` | yes |

Reviewer: ```suggestion
| Output format | Content type           | Streaming? |
|---------------|------------------------|------------|
| JSON          | `application/json`     | No         |
| JSON Lines    | `application/x-ndjson` | Yes        |
| CSV           | `text/csv`             | Yes        |
| Plain text    | `text/plain`           | Yes        |
```

> ## Optimizing Queries

Reviewer: ```suggestion
## Optimizing your queries
```

> All the tables supported by Log Explorer contain a special column called `date`, which helps to narrow down the amount of data that is scanned in order to respond to your query resulting in faster query response times. The value of `date` must be in the form of `YYYY-MM-DD`. For example, to query logs that occurred on October 12, 2023, add the following to your WHERE clause: `date = '2023-10-12'`. The column supports the standard operators of `<`, `>`, `=`.

Reviewer: ```suggestion
All the tables supported by Log Explorer contain a special column called `date`, which helps to narrow down the amount of data that is scanned to respond to your query, resulting in faster query response times. The value of `date` must be in the form of `YYYY-MM-DD`. For example, to query logs that occurred on October 12, 2023, add the following to your `WHERE` clause: `date = '2023-10-12'`. The column supports the standard operators of `<`, `>`, and `=`.
```

```bash

> curl --request GET \
>   --url https://api.cloudflare.com/client/v4/zones/<ZONE_ID>/logs/explorer/query/sql \

Reviewer: ```suggestion
curl https://api.cloudflare.com/client/v4/zones/{zone_id}/logs/explorer/query/sql \
```

  --header 'Authorization: Bearer <API_TOKEN>' \

>   --url-query query="SELECT clientRequestMethod, clientRequestPath, clientRequestProtocol FROM http_requests WHERE date = '2023-10-12' LIMIT 500'"

Reviewer: ```suggestion
  --url-query query="SELECT clientRequestMethod, clientRequestPath, clientRequestProtocol FROM http_requests WHERE date = '2023-10-12' LIMIT 500"
```

```

## FAQs

### Which fields (or columns) are available for querying?

> All fields listed in the datasets [Log Fields](/logs/reference/log-fields/) are viewable in Log Explorer. For filtering, only fields with simple values, such as those of type bool, int, float, or string value are supported. Fields with key-value pairs aren't currently supported. For example, you can't use the fields `RequestHeaders` and `Cookies` from the HTTP Requests dataset in a filter.

Reviewer: ```suggestion
All fields listed in the datasets [Log Fields](/logs/reference/log-fields/) are viewable in Log Explorer. For filtering, only fields with simple values, such as those of type `bool`, `int`, `float`, or `string` are supported. Fields with key-value pairs are currently not supported. For example, you cannot use the fields `RequestHeaders` and `Cookies` from the HTTP requests dataset in a filter.
```

### Why does my query not complete or time out?

Log Explorer performs best when query parameters focus on narrower ranges of time. You may experience query timeouts when your query would return a large quantity of data. Consider refining your query to improve performance.

### My query returned an error. How do I figure out what went wrong?

> We are actively working on improving error codes. If you receive a generic error, check your sql syntax (if you are using the custom sql feature), make sure you have included a date and a limit, and that the field you are filtering is not a key-value pair. If the query still fails it is likely timing out. Try refining your filters. 

Reviewer: ```suggestion
We are actively working on improving error codes. If you receive a generic error, check your SQL syntax (if you are using the custom SQL feature), make sure you have included a date and a limit, and that the field you are filtering is not a key-value pair. If the query still fails it is likely timing out. Try refining your filters. 
```

### Where is the data stored?

The data is stored in Cloudflare R2. Each Log Explorer dataset is stored on a per-customer level, similar to Cloudflare D1, ensuring that your data is kept separate from that of other customers. In the future, this single-tenant storage model will provide you with the flexibility to create your own retention policies and decide in which regions you want to store your data.

> Does Log Explorer support Customer Metadata Boundary?

Reviewer: ```suggestion
### Does Log Explorer support Customer Metadata Boundary?

```
